bin/faucet/src/server/mod.rs

```diff
@@ -0,0 +1,213 @@
+};
+use frontend::Metadata;
+use get_tokens::{GetTokensState, RawMintRequest, get_tokens};
+use http::{HeaderValue, Request};
+use miden_node_utils::grpc::UrlExt;
+use miden_objects::{account::AccountId, block::BlockNumber, note::Note};
+use miden_tx::utils::Serializable;
+use tokio::{net::TcpListener, sync::oneshot};
+use tower::ServiceBuilder;
+use tower_governor::{
+    GovernorError, GovernorLayer,
+    governor::GovernorConfigBuilder,
+    key_extractor::{KeyExtractor, SmartIpKeyExtractor},
+};
+use tower_http::{
+    cors::CorsLayer,
+    set_header::SetResponseHeaderLayer,
+    trace::{DefaultOnResponse, TraceLayer},
+};
+use tracing::Level;
+use url::Url;
+
+use crate::{
+    COMPONENT,
+    faucet::{FaucetId, MintRequest},
+    types::AssetOptions,
+};
+
+mod frontend;
+mod get_tokens;
+
+// FAUCET STATE
+// ================================================================================================
+
+type RequestSender = tokio::sync::mpsc::Sender<(MintRequest, oneshot::Sender<(BlockNumber, Note)>)>;
+
+/// Serves the faucet's website and handles token requests.
+#[derive(Clone)]
+pub struct Server {
+    mint_state: GetTokensState,
+    metadata: &'static Metadata,
+}
+
+impl Server {
+    pub fn new(
+        faucet_id: FaucetId,
+        asset_options: AssetOptions,
+        request_sender: RequestSender,
+    ) -> Self {
+        let mint_state = GetTokensState::new(request_sender, asset_options.clone());
+        let metadata = Metadata {
+            id: faucet_id,
+            asset_amount_options: asset_options,
+        };
+        // SAFETY: Leaking is okay because we want it to live as long as the application.
+        let metadata = Box::leak(Box::new(metadata));
+
+        Server { mint_state, metadata }
+    }
+
+    pub async fn serve(self, url: Url) -> anyhow::Result<()> {
+        // Rate limits by IP. We do additional per account rate limiting in the get_tokens method.
+        //
+        // Limits chosen somewhat arbitrarily, but we have five assets required to load the webpage.
+        // So allowing 8 in a burst seems okay.
+        //
+        // SAFETY: No non-zero elements, so we are okay.
+        let ip_rate_limiter = GovernorConfigBuilder::default()
+            .const_burst_size(8)
+            .const_per_second(1)
+            // The default extractor uses the peer address which is incorrect
+            // if used behind a proxy.
+            .key_extractor(SmartIpKeyExtractor)
+            .finish()
+            .unwrap();
+        let ip_rate_limiter = Arc::new(ip_rate_limiter);
+
+        let account_rate_limiter = GovernorConfigBuilder::default()
+            .const_burst_size(1)
+            .const_per_second(10)
+            // The default extractor uses the peer address which is incorrect
+            // if used behind a proxy.
+            .key_extractor(AccountKeyExtractor)
+            .finish()
+            .unwrap();
+        let account_rate_limiter = Arc::new(account_rate_limiter);
```

Comment:
These rate limit by

- IP address, 8 requests which refill by 1 per second
- account ID, 1 requests, refills after 10s

The latter is only applied to the `get_tokens` endpoint.

To enable this I needed to change how the `get_tokens` works. It no longer takes in json, but rather inputs are via query params.